### ISSUE TITLE
How to train agent to play my own (2048) game?

### ISSUE THREAD

**Issue by gorgitko** [matthiasplappert/keras-rl#38]
Hello,

I am working on AI for the game 2048 (https://gabrielecirulli.github.io/2048/). This library looks like fantastic so I want to ask you what I need to do in order to train agent (some discrete one, probably DQN)?
I have implemented 2048 using NumPy (https://github.com/gorgitko/pygames/tree/master/2048), so I should implement some API for your library? I see in your examples that you are training the OpenAI Gym games so I just need to implement same API in my 2048 game class?

Thank you!

**Comment by matthiasplappert** [comment]
Have a look at the `Env` class (https://github.com/matthiasplappert/keras-rl/blob/master/rl/core.py#L300). What you have to do is implement at least the `step` and `reset` methods. You can then simply pass your environment to the agent and it should work. Depending on your needs, you may wish to have a look at the other methods as well.

**Comment by gorgitko** [comment]
Thanks! I will try it and then it can be used as an example for others who also want to train agent for own custom game.

**Comment by matthiasplappert** [comment]
Sounds great! Let me know if you need anything else.

**Comment by gorgitko** [comment]
So I have implemented the Gym API to my 2048 game class and the learning works :) However, I have a few questions:

1. How to access my environment during training or testing? I would like to print the real game score after each episode. I know I can return `info (dict)` from `step()`, but how to print that on the end of each episode?
2. This is more related to general RL. Reward should be simple or more complex? For example:
    * Simple reward will be `previous_score - new_score` or `-1` when impossible action is performed. But I don't really know if this is a good reward strategy.
    * Complex reward (in 2048 game) won't be computed not only as a single number, but rather as a sum of "mini-rewards" from counts and positions of tiles in current and previous observations. E.g. if you have ever played 2048 game, maybe you noticed that keeping the biggest values in one corner is probably the best strategy, so this should bring a high mini-reward. Next mini-rewards could be computed from the ratio of filled/empty tiles, how many legal steps were performed during current episode, and so on. So should I compute these complex rewards or RL is able to learn the game strategy only from the simple rewards?

Thanks for reply!

**Comment by matthiasplappert** [comment]
Great! Regarding your questions:

1. Depending on what you need, it might also be sufficient to pass `verbose=1` when you call `fit` or `test` to get some stats like the average reward per episode and such. If this is not enough, you can write your own callback. For example, take a look at the `Visualizer` in `callbacks.py`, which interacts with the environment directly. You can also have a look at `TrainIntervalLogger` to get an idea of a more complex callback. Have a look at `Callback` in the same file to see what hooks are available for you to implement.
2. Normalization of the inputs can definitely be helpful. Your scaling seems reasonable but you shouldn't divid by the maximum number on the field (in your example you seem to divid by 4) but rather by the maximum tile number that is possible to achieve (I'm not sure but my guess is that the maximum tile is 2048, thus you should divid by log2(2048) = 11). In short, you would compute `x' = log2(x) / 11` for each tile position. You'll also have to think about how you represent an empty tile in this case. On a related note, it probably makes sense to treat the observation (= the game board) as an image and use 2D convolutions.
3. Design of a good reward function is actually very hard. However, I would suggest to keep it simple and go with option 1 that you described (`previous_score - new_score` or `-1` when impossible action). This makes more sense in my opinion since option 2 would make a lot of assumptions (e.g. it's good to play in a way that is corner-focused) although these assumptions might turn out to be incorrect or suboptimal. Generally speaking, you want to strike a balance between guiding your agent toward a desired goal state but you don't want to design the reward function in a way that is highly biased towards a specific solution that you think could be effective.

I hope that helps and let me know if your agent learns to play the game successfully!

**Comment by gorgitko** [comment]
Thanks!

1. I have `verbose=2`, but of course real game score isn't in the output. I want to do some statistics like counting the tiles etc. So I will inspect the classes you suggest. Btw. maybe it would be good to print the `info` dict returned from `step()` and have a `print_info` parameter in `Agent.fit()`.
2. 
   * Normalizing to maximum **global** value seems reasonable. The game can theoretically continue to tile values like `65536` (`2^16`), but first I will try the `4096` (`2^12`), so `log2 / 12` transform.
   * I don't know very well the convolutional networks. I just know they are used for image recognition, when we first obtain the feature map by trained filtering, then do some pooling and finally use the fully-connected layer. **Has it hypothetically some benefit for this "2048" problem?**
3. Yeah, that makes sense. But still when I think about the simple reward, it seems to me that it's only a short-term reward -- you move in the direction of highest score gain and it's good to raise your score now, but if you move that way, you can mess up current good position of tiles (like the corner-focused composition we spoke about previously). But if I understand it correctly, Q-learning allows to learn the main (long-term) goal from short-term rewards so it learns the rules itself and not from complex rewards we specifically design for it. Feel free to correct me :)

**Comment by matthiasplappert** [comment]
The info dictionary is now supported in the `verbose=1` logger. I'm considering to remove `verbose=2` in the future so it's not implemented for this mode right now. You don't need to change anything, just return an info dictionary from your game environment and it will get logged for you.

**Comment by gorgitko** [comment]
OK, good job :)

One more question: in the `step()` docstring is said "*When end of episode is reached, you are responsible for calling `reset()` to reset this environment's state.*". My code is following:

```
def step(self, action):
    old_score = self.score
    action_status = self.move_insert(action)
    done = self.is_gameover()
    if done:
        self.reset()

    # first false: move is not possible; second false: random tile not inserted
    if action_status == (False, False):
        reward = self.RWD_IMPOSSIBLE_ACTION
    else:
        reward = self.score - old_score

    return self.__normalized_matrix, reward, done, {"score": self.score}
```

I don't know if it makes sense to return the initial board state when is game over. And also what about a reward on game over?

**Comment by matthiasplappert** [comment]
You don't have to call `reset()` yourself in step, the user of the environment needs to take care of this. If you use your environment with Keras-RL, this is taken care of for you automatically. So just remove the if `done: self.reset()` part and you should be good to go.

**Comment by gorgitko** [comment]
Perfect! So maybe just remove that part from docstring.

**Comment by matthiasplappert** [comment]
Good idea, has been removed in https://github.com/matthiasplappert/keras-rl/commit/4180bd1b0d5f5aec1e468decf782024bf28cceb6

**Comment by gorgitko** [comment]
So I am training the agent, but it seems he doesn't make any progress :( After 10 million steps he only learned not to do many invalid steps (relatively to beginning of training), but overall game score is very low.

Here are few points to learning process:

- Board is normalized to `log2` values (so there are powers of two). I think it's better to convert to linear scale than normalize between `0-1` like we discussed [here](https://github.com/matthiasplappert/keras-rl/issues/38#issuecomment-260639578).
- Reward system is following:
  - `new_score - old_score` when valid action is performed (always `> 0`).
  - `p * x`
    - `p` ... invalid step penalty (I set it to `-1`)
    - `x` ... number of current invalid steps which is reset when valid step is performed.
  - So the negative reward is increasing linearly with number of current invalid steps.

Parameters, model topology etc.:

```
# model options
nb_hidden = 256

# memory options
memory_size = int(1e6)
window_length = 100

# training options
nb_steps = int(1e7)
nb_steps_warmup = int(1e4)
nb_actions = 4
target_model_update = 1000
batch_size = 128
train_interval = 100
gamma = 0.90
learning_rate = 1e-4

env = GymBoard()

model = Sequential([
    Flatten(input_shape=(window_length, 4, 4)),
    Dense(nb_hidden),
    Activation("relu"),
    Dense(nb_hidden),
    Activation("relu"),
    Dense(nb_hidden),
    Activation("relu"),
    Dense(nb_actions),
    Activation("linear")
])

memory = SequentialMemory(limit=memory_size, window_length=window_length)
policy = EpsGreedyQPolicy()
dqn = DQNAgent(model=model, nb_actions=nb_actions, memory=memory, nb_steps_warmup=nb_steps_warmup,
               target_model_update=target_model_update, policy=policy, gamma=gamma, batch_size=batch_size,
               train_interval=train_interval, enable_double_dqn=True)

dqn.compile(Adam(lr=learning_rate), metrics=["mae"])
dqn.fit(env, nb_steps=nb_steps, visualize=False, verbose=2, callbacks=[TrainLogger2048()])
```

[And here is a training log.](https://owncloud.cesnet.cz/index.php/s/yUVbkA480u4cETA)

btw. ... The lines in log like
`SCORE: 1440 	VALID STEPS: 298 	INVALID STEPS: 89 	TOTAL STEPS: 387`
are printed by my callback in `on_episode_end()`. These steps are counted directly in env's `move()` method which moves the tiles. But when I look on the next line, there is
`3634931/10000000: episode: 25116, duration: 0.736s, episode steps: 183, steps per second: 249, episode reward: 1400.000, mean reward: 7.650 [-4.000, 128.000], mean action: 1.585 [0.000, 3.000], mean observation: 2.079 [0.000, 7.000], loss: 375.749054, mean_absolute_error: 71.168076, mean_q: 97.146587`
Why there is a different number of steps? Concretely `TOTAL STEPS: 387` vs `episode steps: 183`. When `step()` is called by agent, `move()` must be also called, so `TOTAL STEPS` must be equal to `episode steps`. Or am I missing something?

What is concretely doing `LinearAnnealedPolicy()`? I have a feeling that it is doing the exploration of observation space for a limited number of steps (without actual learning the NN) and then normal policy is used.

Many thanks if you find time for answer!!

**Comment by matthiasplappert** [comment]
- I'm not sure why different number of steps are reported. The code in `TrainEpisodeLogger` seems reasonable to me, so maybe there's a problem in your callback?
- `LinearAnnealedPolicy` wraps another policy and decreases a given parameter linearly. For example, you can use it together with `EpsGreedyQPolicy` to anneal the eps-value from e.g. 1 to 0.1. Also, eps-greedy does *not* mean that the agent first explores and then only exploits. eps-greedy selects the best action (as estimated by the Q function) with probability `1. - eps` and otherwise samples uniformly from the pool of possible actions. In other words, the agent takes a random step with probability eps, in your case with probability 0.1 (default value for eps).

Now, regarding your learning problems:

- First, a window_length of 100 is total overkill. 2048 is fully observable and thus there's really no need for a window length > 1. The reason why a longer window_length may be necessary is that in Atari, for example, the state of the environment is not fully observable if you only look at a single frame.
- Your reward structure may be problematic since it is on an exponential scale. This can be a problem since you're trying to learn a function that is non-stationary and, in later stages of the game, has rewards that are multiple orders of magnitude higher than in early stages. You should try to normalize your step reward in a similar fashion that you use for the observations (i.e. your board representation).
- You could try to clip the delta error to make your algorithm more robust against outliers. In order to do so, you have to set `delta_clip` to a scalar value > 0, e.g. `delta_clip=1.`. However, there was a recent bug that could cause problems with DQN where the error function would so make sure you update keras-rl to the latest available commit first (which is https://github.com/matthiasplappert/keras-rl/commit/0e078a47a49b1a939b6cc2386f96526c6ea346d5)!
- Finally, getting RL algorithms to work is always a lot of work. We are nowhere near a general purpose algorithm yet that can solve any environment. You'll likely have to tune hyperparameters and simple test different things. The above 3 points are, I believe, good starting points.

**Comment by gorgitko** [comment]
Wow, thanks for the superfast answer!

1. To explain it more, my env has two attributes for steps: `n_steps_valid` and `n_steps_invalid`. These are incremented by one when you call `move()`: `n_steps_valid` when valid move is performed, 'n_steps_invalid` otherwise. And `move()` is called in every call of `step()`. In my callback in `on_episode_end()` I just print these attributes, so it must be accurate. But nevermind, this is just a detail :)
2. So `LinearAnnealedPolicy` is like hyper-parameter tuning for policy? Does it make sense to use with my env? Anyway, I will try that.
3. The window_length. OK, I just didn't know what it is really doing.
4. That makes sense, so I will also do the `log2` transform of `new_score - old_score`. Negative reward is not connected with game score so I think transforming it doesn't make sense).
5. I see that `delta_clip=1.` is the default value, so I will keep that.

**Comment by gorgitko** [comment]
So I have done some grid searching of `target_model_update`, `policy` and `gamma` by training models for each of product. Then sometimes when I load weights with `DQNAgent.load_weights()` and do test for one episode with `DQNAgent.test()`, agent is doing so much invalid steps (even over one milion!). But when I do test immediately after training (without loading weights), agent does never do so much invalid steps (almost everytime `n_invalid_steps` < `n_valid_steps`). Is it normal?
Also sometimes training stuck closely to maximum training steps (like on step 1 999 978 from 2 000 000 training steps).

**Comment by gorgitko** [comment]
Anyway, here is my code for 2048 game including environment: https://github.com/gorgitko/MI-MVI_2016
